l2l/optimizers/kalmanfilter/enkf.py

```diff
@@ -0,0 +1,155 @@
+import numpy as np
+import abc
+
+from .kalman_utils import _get_shapes, _encode_targets, _get_batches
+from abc import ABCMeta
+
+try:
+    from numba import jit
+except ModuleNotFoundError:
+    # create a mock decorator if `numba` is not found
+    def jit(**kwargs):
+        def decorator(f):
+            def wrapper(*args, **kwargs):
+                return f(*args, **kwargs)
+            return wrapper
+        return decorator
+
+
+class KalmanFilter(metaclass=ABCMeta):
+    """
+    Abstract Kalman Filter class
+
+    """
+
+    @abc.abstractmethod
+    def fit(self, ensemble, ensemble_size,
+            observations, model_output, gamma, noise, p):
+        """
+        Abstract fit function for the Ensemble Kalman Filter with parameters
+        used in the Ensemble Kalman Filter formulation
+        """
+        pass
+
+
+class EnsembleKalmanFilter(KalmanFilter):
+    def __init__(self, maxit=1, n_batches=1, online=False):
+        """
+        Ensemble Kalman Filter (EnKF)
+
+        EnKF following the formulation found in Iglesias et al. (2013),
+        The Ensemble Kalman Filter for Inverse Problems.
+
+        :param maxit: int, maximum number of iterations
+        :param n_batches, int,  number of batches to used in mini-batch. If set
+            to `1` uses the whole given dataset. Default is `1`.
+        :param online, bool, True if one random data point is requested,
+            between [0, dims], otherwise do mini-batch. `dims` is the number of
+            observations. Default is False
+        """
+        self.Cpp = None
```

Comment:
Usual python convention is to use only lower-case letters for all variable names.